Add Documentation for Integration of Enzyme with Clad (#494)
vgvassilev/clad: docs/userDocs/source/user/UsingEnzymeWithinClad.rst

Using Enzyme within Clad
*************************

> Like Clad, `Enzyme <https://enzyme.mit.edu/>`_ is also a library for Automatic Differentiation(AD) of functions, the only difference being that, Enzyme works at the LLVM IR level, while Clad works at the Clang AST level.

Reviewer: ```suggestion
Like Clad, `Enzyme <https://enzyme.mit.edu/>`_ is also a library for Automatic Differentiation (AD). A major difference is that, Enzyme works at the LLVM IR level, while Clad works at the Clang AST level.
```

> Clad analyses the Clang Abstract Syntax Tree(AST) to generate the derivative of a function, while Enzyme analyses the LLVM IR of a function to generate the derivative of the same.

Reviewer: ```suggestion
Clad analyses the Clang Abstract Syntax Tree (AST) to generate the derivative of a function, while Enzyme analyses the LLVM IR of a function to generate the derivative of the same.
```

AST is used on the previous line but defined here.

> Because multiple languages use the LLVM IR (for example, Rust, C++, Julia, Swift etc.), Enzyme is more interoperable, ie, it supports differentiation of functions written in multiple languages.

Reviewer: ```suggestion
Languages such as Rust, C++, Julia and Swift output LLVM IR which makes Enzyme is more interoperable as it can differentiate functions originally written in multiple languages.
```

> However, in Enzyme, interoperability comes at a cost. For example, while basic functions are supported across various languages; containers, classes and other language specific constructs are seldom supported by Enzyme.

Reviewer: ```suggestion
However, in Enzyme, interoperability comes at a cost. For example, while basic functions are supported across various languages; containers, classes and other language specific constructs require extra scaffolding to be supported by Enzyme.
```

> Clad on the other hand is tightly coupled with the Clang frontend, and therefore with the C++ language, thus giving it complete freedom to address language specific constructs of C++ for Automatic Differentiation.

Reviewer: ```suggestion
Clad is more tightly coupled with the Clang frontend and the C++ language. It has access to the high-level program structure and compile-only constructs such as `consteval`. That allows more coherency when differentiating C++.
```

> The vision of Clad's developers is to enable support for Automatic Differentiation for completely Object Oriented Programs written in C++.

Reviewer: ```suggestion
This includes first class support of Object Oriented Programs written in C++.
```

> The integration of Enzyme within Clad is a feature added to compare both the libraries' performance easily, so that each libraries' pros and cons can be identified faster.

Reviewer: ```suggestion
The initial integration of Enzyme in Clad is to enable cross validation of the produced code.
```

> Currently, use of Enzyme for Reverse Mode AD is supported within clad. This section describes how Enzyme can be used within Clad. 

Reviewer: ```suggestion
Currently, use of Enzyme for Reverse Mode AD is supported within Clad. This section describes how Enzyme can be used within Clad. 
```

Configuring Clad to use Enzyme
=================================

> To enable the use of enzyme within clad, one needs to configure Clad to use Enzyme. This can be done by adding the flag ``-DENABLE_ENZYME_BACKEND=On`` to cmake while configuring clad build.

Reviewer: ```suggestion
To enable the use of enzyme within Clad, one needs to configure Clad to use Enzyme. This can be done by adding the flag ``-DENABLE_ENZYME_BACKEND=On`` to cmake while configuring Clad build.
```

We should spell out Clad consistently.

> Thus the overall cmake command must look something like this

Reviewer: ```suggestion
Thus the overall cmake command should look something like this:
```

.. code-block:: bash

    cmake ../clad -DClang_DIR=/usr/lib/llvm-11 -DLLVM_DIR=/usr/lib/llvm-11 -DCMAKE_INSTALL_PREFIX=../inst -DLLVM_EXTERNAL_LIT="``which lit``" -DENABLE_ENZYME_BACKEND=On

> This flag clones the enzyme repository, builds it, runs tests on it and links it as a static library to the clad shared object during the build process.

Reviewer: ```suggestion
This flag instruct the build system to download and build Enzyme. Then it is linked as a static library to the Clad.
```

Asking clad to generate gradients with Enzyme
================================================

The following code snippet shows how one can request clad to generate gradients with Enzyme::

    #include "clad/Differentiator/Differentiator.h"

    double foo(double* arr) { return arr[0] * arr[1]; }

    int main(){
        auto grad = clad::gradient<clad::opts::use_enzyme>(f1);
        double v[2] = {3, 4};
        double g[2] = {0};
        grad.execute(v, g);
        printf("d_x = %.2f, d_y = %.2f\n", g[0], g[1]); 
    }

Thus, the calling convention is to use ``clad::gradient<clad::opts::use_enzyme>(...)`` instead of the usual calling convention, ``clad::gradient(...)``.

> Calling ``execute``, ``dump`` and other functionalities remain same as that of clad.

Reviewer: What happens if we call `dump` on a enzyme-produced gradient?

Author: It outputs the wrapper function that contains the call to `__enzyme__autodiff`

Extent of support for Enzyme within Clad
=========================================

Currently functions that take in arrays, pointers and primitive types(integer and real) as parameters are supported for differentiation with Enzyme within Clad.
For more ideas on the type of functions supported, please have a look at ``test/Enzyme/ReverseMode.C`` for examples that can be differentiated within Clad.

> 

Reviewer: Can you remove the these stray lines?

> For more ideas on the type of functions supported, have a look at ``test/Enzyme/ReverseMode.C`` for examples that can be differentiated within clad.

Reviewer: ```suggestion
For more ideas on the type of functions supported, please have a look at ``test/Enzyme/ReverseMode.C`` for examples that can be differentiated within clad.
```
